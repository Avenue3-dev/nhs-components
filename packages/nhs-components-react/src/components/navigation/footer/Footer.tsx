import Container, {
  ContainerProps,
} from '@/components/core/container/Container'
import { HTMLProps } from 'react'
import clsx from 'clsx'
import { AsElementLink } from '@/types/link-like'

type FooterProps = {
  visuallyHiddenText?: string
  containerProps?: ContainerProps
  footerProps?: HTMLProps<HTMLDivElement>
} & HTMLProps<HTMLDivElement>

type ListItemProps = {
  containerProps?: HTMLProps<HTMLLIElement>
} & AsElementLink<HTMLAnchorElement>

type Footer = {
  List: typeof List
  ListItem: typeof ListItem
  Copyright: typeof Copyright
} & React.FC<FooterProps>

const List: React.FC<HTMLProps<HTMLUListElement>> = ({
  children,
  className,
  ...rest
}) => {
  return (
    <ul className={clsx('nhsuk-footer__list', className)} {...rest}>
      {children}
    </ul>
  )
}

const ListItem: React.FC<ListItemProps> = ({
  children,
  className,
  asElement: Component = 'a',
  containerProps = {},
  ...rest
}) => {
  const { className: liClassName, ...containerRest } = containerProps

  return (
    <li
      className={clsx('nhsuk-footer__list-item', liClassName)}
      {...containerRest}
    >
      <Component
        className={clsx('nhsuk-footer__list-item-link', className)}
        {...rest}
      >
        {children}
      </Component>
    </li>
  )
}

const Copyright: React.FC<HTMLProps<HTMLParagraphElement>> = ({
  children,
  className,
  ...rest
}) => {
  return (
    <p className={clsx('nhsuk-footer__copyright', className)} {...rest}>
      {children}
    </p>
  )
}

const Footer: Footer = ({
  children,
  className,
  containerProps,
  footerProps = { role: 'contentinfo' },
  visuallyHiddenText = 'Support links',
  ...rest
}): JSX.Element => {
  return (
    <footer {...{ role: 'contentinfo', ...footerProps }}>
      <div className={clsx('nhsuk-footer', className)} {...rest}>
        <Container {...containerProps}>
          {visuallyHiddenText && (
            <h2 className="nhsuk-u-visually-hidden">{visuallyHiddenText}</h2>
          )}
          {children}
        </Container>
      </div>
    </footer>
  )
}

Footer.displayName = 'Footer'
List.displayName = 'Footer.List'
ListItem.displayName = 'Footer.ListItem'

Footer.List = List
Footer.ListItem = ListItem
Footer.Copyright = Copyright

export default Footer
