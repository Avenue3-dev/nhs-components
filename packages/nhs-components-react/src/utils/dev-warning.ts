const prod = process.env.NODE_ENV === 'production'

export const experimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  if (prod) {
    return
  }

  console.warn(`ATTENTION
  This component is tagged as experimental and therefore could change without notice. See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}
