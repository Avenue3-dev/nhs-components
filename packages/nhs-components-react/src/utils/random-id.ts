export const generateRandomString = (length = 5): string => {
  const rand = Math.random() + 1

  return rand.toString(36).substring(2, length + 2)
}

export const generateRandomName = (prefix?: string): string => {
  const randString = generateRandomString()

  return prefix ? `${prefix}_${randString}` : randString
}

export const generateRandomId = generateRandomName
