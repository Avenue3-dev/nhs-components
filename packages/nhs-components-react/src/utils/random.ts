export const generateRandomString = (length: number = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
