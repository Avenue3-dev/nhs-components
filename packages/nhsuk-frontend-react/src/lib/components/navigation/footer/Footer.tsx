import Container, {
  ContainerProps,
} from '@/components/core/container/Container'
import { HTMLProps } from 'react'
import clsx from 'clsx'
import { AsElementLink } from '@/types/link-like'
import { VisuallyHiddenProps } from '@/types/visually-hidden'

type FooterProps = {
  containerProps?: ContainerProps
  footerContainerProps?: HTMLProps<HTMLDivElement>
  footerProps?: HTMLProps<HTMLDivElement>
} & HTMLProps<HTMLDivElement> &
  Partial<VisuallyHiddenProps>

type ListItemProps = {
  containerProps?: HTMLProps<HTMLLIElement>
} & AsElementLink<HTMLAnchorElement>

type Footer = {
  List: typeof List
  ListItem: typeof ListItem
  Copyright: typeof Copyright
} & React.FC<FooterProps>

const List: React.FC<HTMLProps<HTMLUListElement>> = ({
  children,
  className,
  ...rest
}) => {
  return (
    <ul className={clsx('nhsuk-footer__list', className)} {...rest}>
      {children}
    </ul>
  )
}

const ListItem: React.FC<ListItemProps> = ({
  children,
  className,
  asElement: Component = 'a',
  containerProps = {},
  ...rest
}) => {
  const { className: liClassName, ...containerRest } = containerProps

  return (
    <li
      className={clsx('nhsuk-footer__list-item', liClassName)}
      {...containerRest}
    >
      <Component
        className={clsx('nhsuk-footer__list-item-link', className)}
        {...rest}
      >
        {children}
      </Component>
    </li>
  )
}

const Copyright: React.FC<HTMLProps<HTMLParagraphElement>> = ({
  children,
  className,
  ...rest
}) => {
  return (
    <p className={clsx('nhsuk-footer__copyright', className)} {...rest}>
      {children}
    </p>
  )
}

const Footer: Footer = ({
  children,
  className,
  containerProps,
  footerContainerProps,
  footerProps = { role: 'contentinfo' },
  visuallyHiddenText = 'Support links',
  ...rest
}): JSX.Element => {
  const { className: footerContainerClassName, ...footerContainerRest } = footerContainerProps || {}

  return (
    <footer {...{ role: 'contentinfo', ...footerProps }}>
      <div className={clsx('nhsuk-footer-container', footerContainerClassName)} {...footerContainerRest}>
        <Container {...containerProps}>
          <div className={clsx('nhsuk-footer', className)} {...rest}>
              {visuallyHiddenText && (
                <h2 className="nhsuk-u-visually-hidden">{visuallyHiddenText}</h2>
              )}
              {children}
          </div>
        </Container>
      </div>
    </footer>
  )
}

Footer.displayName = 'Footer'
List.displayName = 'Footer.List'
ListItem.displayName = 'Footer.ListItem'
Copyright.displayName = 'Footer.Copyright'

Footer.List = List
Footer.ListItem = ListItem
Footer.Copyright = Copyright

export default Footer
