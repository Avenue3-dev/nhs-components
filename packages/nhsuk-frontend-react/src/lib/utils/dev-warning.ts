const prod = process.env.NODE_ENV === 'production'

/**
 * Print a warning when using experimental components when not in production mode.
 *
 * @param component name of the experimental component.
 */
export const useExperimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  useDevWarning(`This component is tagged as experimental and therefore could change without notice.

  See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}

/**
 * Print a warning message when not in production mode.
 *
 * @param message message to print when not it production.
 */
export const useDevWarning = (message: string): void => {
  if (prod) {
    return
  }

  console.warn(message)
}
