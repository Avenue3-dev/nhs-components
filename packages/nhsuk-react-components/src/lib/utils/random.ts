export const generateRandomString = (length = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
