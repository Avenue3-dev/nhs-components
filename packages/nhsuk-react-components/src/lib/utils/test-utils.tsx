/* eslint-disable import/export */
import {
  RenderOptions,
  RenderResult,
  cleanup,
  render,
} from '@testing-library/react'
import { MemoryRouter as Router } from 'react-router-dom'
import { afterEach } from 'vitest'

afterEach(() => {
  cleanup()
})

const customRender = (
  ui: React.ReactElement,
  options: RenderOptions = {},
): RenderResult => {
  return render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => <Router>{children}</Router>,
    ...options,
  })
}

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
// override render export
export { customRender as render }
