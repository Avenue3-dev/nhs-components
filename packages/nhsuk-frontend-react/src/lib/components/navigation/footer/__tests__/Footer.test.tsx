import { render, screen } from '@/utils/test-utils'
import { describe, expect, test } from 'vitest'
import Footer from '..'

describe('Footer', () => {
  test('should match the snapshot', () => {
    render(
      <Footer data-testid="footer">
        <Footer.List>
          <Footer.ListItem href="#accessibility">
            Accessibility statement
          </Footer.ListItem>
          <Footer.ListItem href="#contact-us">Contact us</Footer.ListItem>
        </Footer.List>
        <Footer.Copyright>&copy; Crown copyright</Footer.Copyright>
      </Footer>,
    )

    expect(screen.getByTestId('footer').parentNode).toMatchSnapshot()
  })
})
