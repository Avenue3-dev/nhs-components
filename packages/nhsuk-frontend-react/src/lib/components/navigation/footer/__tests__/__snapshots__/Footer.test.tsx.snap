// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Footer > should match the snapshot 1`] = `
<footer
  role="contentinfo"
>
  <div
    class="nhsuk-footer"
    data-testid="footer"
  >
    <div
      class="nhsuk-width-container"
    >
      <h2
        class="nhsuk-u-visually-hidden"
      >
        Support links
      </h2>
      <ul
        class="nhsuk-footer__list"
      >
        <li
          class="nhsuk-footer__list-item"
        >
          <a
            class="nhsuk-footer__list-item-link"
            href="#accessibility"
          >
            Accessibility statement
          </a>
        </li>
        <li
          class="nhsuk-footer__list-item"
        >
          <a
            class="nhsuk-footer__list-item-link"
            href="#contact-us"
          >
            Contact us
          </a>
        </li>
      </ul>
      <p
        class="nhsuk-footer__copyright"
      >
        Â© Crown copyright
      </p>
    </div>
  </div>
</footer>
`;
